### PR TITLE
Integrate new ignored loss in semantic segmentation

### PR DESCRIPTION
**Pull request by supersoob**
### Summary

This PR includes new ignored cross entropy loss for semantic segmentation.

**Optimization for segmentation**
- [x] (DONE) sampler change
- [x] (DONE) MaskCompose remove
- [x] (DONE) repeated ckpt loading remove
- [x] (In review) Ignored loss function - **this PR is handling this**
- [ ] (TODO) refactoring + unit test
- [ ] (TODO) head aggregator change

**Results**
image

### Checklist

- [x] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by eunwoosh** [comment]
Could you change the PR as a draft if it's in progress?

**Comment by supersoob** [comment]
I'm ready to review. Let me add related unit tests in other PR.

**Review by JihwanEom — DISMISSED** [review]

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py [hunk 1]
```diff
@@ -4,30 +4,138 @@
...
+from otx.algorithms.segmentation.adapters.mmseg.utils import (
+    get_valid_label_mask_per_batch,
 )
 
 
 @HEADS.register_module()
-class CustomFCNHead(
-    SegmentOutNormMixin, AggregatorMixin, MixLossMixin, PixelWeightsMixin2, FCNHead
-):  # pylint: disable=too-many-ancestors
+class CustomFCNHead(FCNHead):  # pylint: disable=too-many-ancestors
```
Comment (jaegukhyun) [inline otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py]: Could we remove this pylint?

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py [hunk 2]
```diff
@@ -4,30 +4,138 @@
...
     """Custom Fully Convolution Networks for Semantic Segmentation."""
 
-    def __init__(self, *args, **kwargs):
+    def __init__(
+        self,
+        enable_aggregator=False,
```
Comment (jaegukhyun) [inline otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py]: How about adding type hint for this class?

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py [hunk 3]
```diff
@@ -4,30 +4,138 @@
...
+
+    def _transform_inputs(self, inputs):
+        if self.aggregator is not None:
+            inputs = self.aggregator(inputs)[0]
+        else:
+            inputs = super()._transform_inputs(inputs)
+
+        return inputs
+
+    def forward_train(self, inputs, img_metas, gt_semantic_seg, train_cfg):
```
Comment (jaegukhyun) [inline otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py]: It would be much better if all functions have own type hint!

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py [hunk 4]
```diff
@@ -6,58 +6,60 @@
...
-class CrossEntropyLossWithIgnore(OTXBasePixelLoss):
-    """CrossEntropyLossWithIgnore with Ignore Mode Support for Class Incremental Learning.
+class CrossEntropyLossWithIgnore(CrossEntropyLoss):
+    """CrossEntropyLossWithIgnore with Ignore Mode Support for Class Incremental Learning."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._loss_name = "loss_ce_ignore"
+
+    def forward(
```
Comment (jaegukhyun) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py:29]: Could you add type-hints here?

**Review by jaegukhyun — COMMENTED** [review]
Overall codes look good to me. Thank you for your effort. Could you check my comments. I want you to add type-hint for new classes.

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py [hunk 5]
```diff
@@ -4,30 +4,143 @@
...
+from otx.algorithms.segmentation.adapters.mmseg.utils import (
+    get_valid_label_mask_per_batch,
 )
 
 
 @HEADS.register_module()
-class CustomFCNHead(
-    SegmentOutNormMixin, AggregatorMixin, MixLossMixin, PixelWeightsMixin2, FCNHead
-):  # pylint: disable=too-many-ancestors
+class CustomFCNHead(FCNHead):
```
Comment (sungmanc) [inline otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py:24]: For this class, more detailed docstring is needed for users since it has many attributes and each of them is hard to understand.

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py (hunk 1 above)
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py]: Thanks for comments. I removed.

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py (hunk 2 above)
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py]: Thanks for comments. I added.

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py (hunk 3 above)
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py]: Thanks for comments. I added.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py (hunk 4 above)
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py:29]: Thanks for comments. I added.

**Review by jaegukhyun — DISMISSED** [review]

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py [hunk 6]
```diff
@@ -3,61 +3,65 @@
...
+        cls_score: Optional[torch.Tensor],
+        label: Optional[torch.Tensor],
+        weight: Optional[torch.Tensor] = None,
+        avg_factor: Optional[int] = None,
+        reduction_override: Optional[str] = "mean",
+        ignore_index: int = 255,
+        valid_label_mask: Optional[torch.Tensor] = None,
+        **kwargs
+    ):
+        """Forward."""
```
Comment (sungmanc) [inline otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py]: Same with above, to understand this function, more detailed docstring is needed.
I couldn't totally understand avg_factor, valid_label_mask.

File: otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py (hunk 5 above)
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/heads/custom_fcn_head.py:24]: I updated. it was hard to investigate its code that was not written any docs in the original repo.

File: otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py (hunk 6 above)
Comment (supersoob) [reply otx/algorithms/segmentation/adapters/mmseg/models/losses/cross_entropy_loss_with_ignore.py]: I updated, and avg_factor used in default args in cr loss.

**Review by jaegukhyun — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]
Thanks a lot for your hard work and quick implmentation :)

**Review by goodsong81 — APPROVED** [review]
LGTM! The test cases look enough for the sanity check. Thank you :)

**Review by sungmanc — APPROVED** [review]
